pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,331 @@
+package pxc
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"github.com/percona/percona-xtradb-cluster-operator/pkg/pxc/users"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const internalPrefix = "internal-"
+
+func (r *ReconcilePerconaXtraDBCluster) reconcileUsers(cr *api.PerconaXtraDBCluster) error {
+	if cr.Status.Status != api.AppStateReady {
+		return nil
+	}
+
+	sysUsersSecretObj := corev1.Secret{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
+		},
+		&sysUsersSecretObj,
+	)
+	if err != nil && k8serrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrapf(err, "get sys users secret '%s'", cr.Spec.SecretsName)
+	}
+
+	newSysData, err := json.Marshal(sysUsersSecretObj.Data)
+	if err != nil {
+		return errors.Wrap(err, "marshal sys secret data")
+	}
+	newSecretDataHash := sha256Hash(newSysData)
+
+	internalSysSecretObj, err := r.getInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "get internal sys users secret")
+	}
+
+	dataChanged, err := sysUsersSecretDataChanged(newSecretDataHash, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "check sys users data changes")
+	}
+
+	if !dataChanged {
+		return nil
+	}
+
+	restartPXC, restartProxy, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	err = r.updateInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
```

Comment:
1) should we also check `cr.Spec.ProxySQL.Enabled`?
2) we just update hash, is it really needed to check size?
```suggestion
	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Enabled {
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/418/commits/b26d5047e5d7ab5ca27cda071d72279d01ba49fa